### PR TITLE
logs: add support for logs processing with the SDS shared library.

### PR DESCRIPTION
**Pull request by remeh**
APR-16

This pull-request:
  * adapts the log processing pipeline to support the SDS shared library for sensitive data processing
  * subscribes to 2 new Remote Configuration products to receive configurations for the new logs processing support
  * adds support for "processing tags" in the logs pipeline: at processing time, the processor can accumulate tags that are added later one to the event

The feature uses a `sds` Go build flag: if not set, the Agent compiles against a mocked implementation not using the shared library.

### Additional Notes

The diff is a bit large but most of the new lines are either isolated in the new package, or tests/go.sum: changes to the existing part of the Agent are limited and so is the impact. That's also a reason I did not add the `major_change` label: it does not impact multiple teams a lot and it is not changing many things from the existing behaviour of the Logs Agent.

A following PR will adapt the build pipeline to compile the shared library, to run SDS unit tests (they need the shared library), compile the Agent linked to the shared library, and more things related to the build pipeline.

### Describe how to test/QA your changes

* Run the Agent in an org where this feature is enabled.
* Go through the SDS configuration page / Agent and configure/reconfigure rules, make sure looking at the Agent logs that it is receiving the configuration updates and properly applying them
* Collect logs matching one or several of these rules and validate the configuration actions are applied.

### CODE REVIEW

File: pkg/logs/processor/processor.go [hunk 1]
```diff
@@ -82,17 +96,31 @@ func (p *Processor) run() {
...
-		p.mu.Unlock()
+
+	for {
+		select {
+		case msg, ok := <-p.inputChan:
+			if !ok { // channel has been closed
+				return
+			}
+			p.processMessage(msg)
+			p.mu.Lock() // block here if we're trying to flush synchronously
```
Comment (ameske) [inline pkg/logs/processor/processor.go:107]: Just curious, as I see this was already in the implementation: but what's being synchronized here since this is a back to back Lock/Unlock call.

File: comp/logs/agent/agent.go [hunk 2]
```diff
@@ -278,3 +290,57 @@ func (a *agent) GetMessageReceiver() *diagnostic.BufferedMessageReceiver {
...
+	}
+
+	// Apply the new status to all configs
+	for cfgPath := range updates {
+		if err == nil {
+			applyStateCallback(cfgPath, state.ApplyStatus{State: state.ApplyStateAcknowledged})
+		} else {
+			applyStateCallback(cfgPath, state.ApplyStatus{
+				State: state.ApplyStateError,
+				Error: err.Error(),
```
Comment (ameske) [inline comp/logs/agent/agent.go:314]: Just making sure that the error string here is something you'd be ok with being visible to users.

**Review by ameske — APPROVED** [review]

File: pkg/logs/processor/processor.go (hunk 1 above)
Comment (remeh) [reply pkg/logs/processor/processor.go:107]: [It's a change I did long in the past](https://github.com/DataDog/datadog-agent/commit/519c2cca1ed58219887cfde95d5a7aaed5f59373#diff-575596a24724456a569a4efcf00ccb19ec24b660ac3b6707de509e3c6f8b0b91R52-R91) but IIRC, it's because it is OK to finish the processing/call to processMessage, but we don't want to read one more message from inputChan if we've concurrently started to do a synchronous call to Flush. Rephrased, having this Lock/Unlock at this level freezes this async for loop from reading inputChan while inputChan is "synchronously" drained in Flush. If that makes sense.

File: pkg/logs/sds/scanner.go [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+			return nil
+		}
+		return nil
+	}
+
+	// prepare the scanner rules
+	var sdsRules []sds.Rule
+	for _, userRule := range config.Rules {
+		// read the rule in the standard rules
+		standardRule := s.standardRules.GetByID(userRule.Definition.StandardRuleID)
```
[the remark this answers is not in the document]
Comment (remeh) [reply pkg/logs/sds/scanner.go:160]: Hundreds, worse case thousands, I don't think it matters much since it's only when reconfiguring. But still it won't be a big effort to use a map instead, I'll change this. Good catch thank you.

File: pkg/logs/sds/scanner.go [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+
+	// store the raw configuration for a later refresh
+	// if we receive new standard rules
+	s.rawConfig = rawConfig
+	s.configuredRules = config.Rules
+
+	telemetry.GetStatsTelemetryProvider().Count(tlmSDSConfigurationSuccess, 1, []string{})
+	telemetry.GetStatsTelemetryProvider().Count(tlmSDSConfiguredRules, float64(len(s.configuredRules)), []string{})
+
+	log.Info("Created an SDS scanner with", len(scanner.Rules), "rules")
```
[the remark this answers is not in the document]
Comment (remeh) [reply pkg/logs/sds/scanner.go:221]: Good catch, too used to stdlib `fmt`/`log` behaviour.

File: pkg/logs/sds/scanner.go [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+		default:
+			log.Warnf("Unknown MatchAction type (%v) for rule '%s':", matchAction, standardRule.Name)
+		}
+	}
+
+	// create the new SDS Scanner
+	var scanner *sds.Scanner
+	var err error
+	if scanner, err = sds.CreateScanner(sdsRules); err != nil {
+		telemetry.GetStatsTelemetryProvider().Count(tlmSDSConfigurationError, 1, []string{"error:cant_create_scanner"})
```
[the remark this answers is not in the document]
Comment (remeh) [reply pkg/logs/sds/scanner.go:203]: Let me remove these completely for the time being. What/how we want to report is still in discussion.

File: pkg/logs/pipeline/provider.go [hunk 6]
```diff
@@ -106,6 +110,31 @@ func (p *provider) Stop() {
 	p.outputChan = nil
 }
 
+func (p *provider) ReconfigureSDS(standardRules []byte, rules []byte) error {
+	var order sds.ReconfigureOrder
+
+	if standardRules != nil {
```
[the remark this answers is not in the document]
Comment (remeh) [reply pkg/logs/pipeline/provider.go:116]: Good point.

File: pkg/logs/sds/scanner.go [hunk 7]
```diff
@@ -0,0 +1,295 @@
...
+// This is NOT reconfiguring the internal SDS scanner, call `reconfigureRules`
+// if you have to.
+// This method is NOT thread safe, the caller has to ensure the thread safety.
+func (s *Scanner) reconfigureStandardRules(rawConfig []byte) error {
+	if rawConfig == nil {
+		return fmt.Errorf("Invalid nil raw configuration for standard rules")
+	}
+
+	var standardRules StandardRulesConfig
+	if err := json.Unmarshal(rawConfig, &standardRules); err != nil {
```
[the remark this answers is not in the document]
Comment (remeh) [reply pkg/logs/sds/scanner.go:105]: It's a good idea, l'll look into this. (the number of pipelines is the number of logs agent pipelines -> 4 by default)

[the remark this answers is not in the document]
Comment (remeh) [reply pkg/logs/sds/scanner.go:105]: Unfortunately I can't with current design: the types are not accessible before propagating to pipelines (it would create an import cycle).
If you don't mind, I'll keep the unmarshal per pipeline for now, as it is only on reconfiguration, it should be OK.

File: pkg/logs/pipeline/provider.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (remeh) [reply pkg/logs/pipeline/provider.go:116]: Should be addressed.

File: pkg/logs/sds/scanner.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (remeh) [reply pkg/logs/sds/scanner.go:160]: Addressed.

File: pkg/logs/sds/scanner.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (remeh) [reply pkg/logs/sds/scanner.go:221]: Done!

File: pkg/logs/sds/scanner.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (remeh) [reply pkg/logs/sds/scanner.go:203]: Addressed.

File: pkg/logs/sds/scanner_test.go [hunk 8]
```diff
@@ -232,6 +232,10 @@ func TestCreateScanner(t *testing.T) {
 		Config: agentConfig,
 	})
 
+	if err != nil {
+		t.Error("no error should happen")
+	}
```
Comment (vickenty) [inline pkg/logs/sds/scanner_test.go:237]: Would it make sense to use `require.NoError` here?

File: comp/logs/agent/agent_core_init.go [hunk 9]
```diff
@@ -64,7 +64,6 @@ func (a *agent) SetupPipeline(processingRules []*config.ProcessingRule, wmeta op
 	a.launchers = lnchrs
 	a.health = health
 	a.diagnosticMessageReceiver = diagnosticMessageReceiver
-
```
Comment (vickenty) [inline comp/logs/agent/agent_core_init.go:67]: Would it be possible to avoid changes to unrelated files?

File: comp/logs/agent/config/integration_config.go [hunk 10]
```diff
@@ -261,6 +261,7 @@ func (c *LogsConfig) Validate() error {
 	if err != nil {
 		return err
 	}
+
```
Comment (vickenty) [inline comp/logs/agent/config/integration_config.go:264]: Same here.

File: pkg/logs/pipeline/provider.go [hunk 11]
```diff
@@ -106,6 +110,53 @@ func (p *provider) Stop() {
 	p.outputChan = nil
 }
 
+func (p *provider) reconfigureSDS(config []byte, agentConfig bool) error {
+	var responses []chan error
+
+	typ := sds.StandardRules
+	if agentConfig {
+		typ = sds.AgentConfig
+	}
```
Comment (vickenty) [inline pkg/logs/pipeline/provider.go:119]: Would it make sense to accept `typ` directly? This function is called from the same file, callers have access to the same constants.

File: pkg/logs/processor/processor.go [hunk 12]
```diff
@@ -82,17 +96,34 @@ func (p *Processor) run() {
...
+			//nolint:staticcheck
+			p.mu.Unlock()
+		case order := <-p.ReconfigChan:
+			p.mu.Lock()
+			if err := p.sds.Reconfigure(order); err != nil {
+				log.Error("Error while reconfiguring the SDS scanner:", err)
```
Comment (vickenty) [inline pkg/logs/processor/processor.go:113]:
```suggestion
				log.Errorf("Error while reconfiguring the SDS scanner: %v", err)
```

**Review by vickenty — APPROVED** [review]

File: pkg/logs/sds/scanner_test.go (hunk 8 above)
Comment (remeh) [reply pkg/logs/sds/scanner_test.go]: Converted all new unit tests to use `testify.Require`

File: comp/logs/agent/config/integration_config.go (hunk 10 above)
Comment (remeh) [reply comp/logs/agent/config/integration_config.go]: Yup, sorry.

File: pkg/logs/pipeline/provider.go (hunk 11 above)
Comment (remeh) [reply pkg/logs/pipeline/provider.go]: You're right 🤦 thanks.

File: pkg/logs/processor/processor.go (hunk 12 above)
Comment (remeh) [reply pkg/logs/processor/processor.go:113]: Done

File: pkg/logs/processor/processor.go [hunk 13]
```diff
@@ -141,7 +175,18 @@ func (p *Processor) applyRedactingRules(msg *message.Message) bool {
 		}
 	}
 
-	// TODO(remy): this is most likely where we want to plug in SDS
+	// Use the SDS implementation
+	// --------------------------
+
+	// Global SDS scanner, applied on all log sources
+	if p.sds.IsReady() {
```
Comment (gh123man) [inline pkg/logs/processor/processor.go:182]: The way I read this, it seems like it is possble for logs to flow through the processor before sds is ready. 
Is that correct (and is that what we want to do)? 

I may be lacking some context here but I would assume if SDS is enabled, we do not want any logs to bypass it

**Review by mackjmr — APPROVED** [review]

**Review by gh123man — APPROVED** [review]

File: comp/logs/agent/agent.go [hunk 14]
```diff
@@ -282,3 +296,55 @@ func (a *agent) GetMessageReceiver() *diagnostic.BufferedMessageReceiver {
...
+			})
+		}
+	}
```
Comment (chouetz) [inline comp/logs/agent/agent.go:349]: As this block is a dupe of lines 312-322 you could factorise

File: pkg/logs/sds/rules.go [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+	Description             string   `json:"description"`
+	Pattern                 string   `json:"pattern"`
+	Tags                    []string `json:"tags"`
+	Type                    string   `json:"type"`
+	DefaultIncludedKeywords []string `json:"default_included_keywords"`
+}
+
+// StandardRulesConfig contains standard rules.
+type StandardRulesConfig struct {
+	Rules []StandardRuleConfig `json:"rules"`
```
Comment (chouetz) [inline pkg/logs/sds/rules.go:40]:
```suggestion
	Rules []StandardRuleConfig `json:"standard_rules"`
```
this is maybe not necessary I don't have the full json format to see :)

**Review by chouetz — APPROVED** [review]
Minor comments

**Review by pducolin — APPROVED** [review]

**Comment by pducolin** [comment]
Ok for @DataDog/agent-developer-tools owned files (tasks)

**Review by rtrieu — APPROVED** [review]
LGTM
